packages/react-reconciler/src/ReactFiberRootScheduler.js

```diff
@@ -0,0 +1,458 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import type {FiberRoot} from './ReactInternalTypes';
+import type {Lane} from './ReactFiberLane';
+import type {PriorityLevel} from 'scheduler/src/SchedulerPriorities';
+
+import {
+  NoLane,
+  NoLanes,
+  getHighestPriorityLane,
+  getNextLanes,
+  includesOnlyNonUrgentLanes,
+  includesSyncLane,
+  markStarvedLanesAsExpired,
+} from './ReactFiberLane';
+import {
+  CommitContext,
+  NoContext,
+  RenderContext,
+  getExecutionContext,
+  getWorkInProgressRoot,
+  getWorkInProgressRootRenderLanes,
+  isWorkLoopSuspendedOnData,
+  performConcurrentWorkOnRoot,
+  performSyncWorkOnRoot,
+} from './ReactFiberWorkLoop';
+import {LegacyRoot} from './ReactRootTags';
+import {
+  ImmediatePriority as ImmediateSchedulerPriority,
+  UserBlockingPriority as UserBlockingSchedulerPriority,
+  NormalPriority as NormalSchedulerPriority,
+  IdlePriority as IdleSchedulerPriority,
+  cancelCallback as Scheduler_cancelCallback,
+  scheduleCallback as Scheduler_scheduleCallback,
+  now,
+} from './Scheduler';
+import {
+  DiscreteEventPriority,
+  ContinuousEventPriority,
+  DefaultEventPriority,
+  IdleEventPriority,
+  lanesToEventPriority,
+} from './ReactEventPriorities';
+import {supportsMicrotasks, scheduleMicrotask} from './ReactFiberHostConfig';
+
+import ReactSharedInternals from 'shared/ReactSharedInternals';
+const {ReactCurrentActQueue} = ReactSharedInternals;
+
+// A linked list of all the roots with pending work. In an idiomatic app,
+// there's only a single root, but we do support multi root apps, hence this
+// extra complexity. But this module is optimized for the single root case.
+let firstScheduledRoot: FiberRoot | null = null;
+let lastScheduledRoot: FiberRoot | null = null;
+
+// Used to prevent redundant mircotasks from being scheduled.
+let didScheduleMicrotask: boolean = false;
+// `act` "microtasks" are scheduled on the `act` queue instead of an actual
+// microtask, so we have to dedupe those separately. This wouldn't be an issue
+// if we required all `act` calls to be awaited, which we might in the future.
+let didScheduleMicrotask_act: boolean = false;
+
+// Used to quickly bail out of flushSync if there's no sync work to do.
+let mightHavePendingSyncWork: boolean = false;
+
+let isFlushingWork: boolean = false;
+
+export function ensureRootIsScheduled(root: FiberRoot): void {
```

Comment:
This looks like a bunch of new code but it's less new than it looks.

This module replaces the `ReactFiberSyncTaskQueue` module that existed previously. Instead of being a generic callback queue, we now maintain a list of all the FiberRoots that have work schedule on them. This was effectively what we were already doing because the only type of callback that was ever scheduled was `performSyncWorkOnWork`.

Then, since we have a list of roots with pending work, I also use that same data structure to stash the root until the microtask where we can do our scheduling logic. So I moved `ensureRootIsScheduled` (where most of that logic lives) into this file to try to keep it all colocated.

I originally was just going to dump this all into the work loop module but there's enough going on here that I think it merits its own module. The way I think of it is that the work loop contains all the logic for a particular render + commit phase, and this module is about scheduling those render cycles with the host environment.

Reply:
The best way to review this PR is probably to read this module from top to bottom.